hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -0,0 +1,273 @@
+    val sparkEngine = new HoodieSparkEngineContext(new JavaSparkContext(spark.sparkContext))
+    val properties = new Properties()
+    properties.putAll(options.asJava)
+    val metadataConfig = HoodieMetadataConfig.newBuilder.fromProperties(properties).build()
+
+    val queryPartitionPath = FSUtils.getRelativePartitionPath(new Path(basePath), queryPath)
+    // Load all the partition path from the basePath, and filter by the query partition path.
+    val partitionPaths = FSUtils.getAllPartitionPaths(sparkEngine, metadataConfig, basePath).asScala
+      .filter(_.startsWith(queryPartitionPath))
+
+    val maxListParallelism = options.get(HoodieWriteConfig.MAX_LISTING_PARALLELISM)
+      .map(_.toInt).getOrElse(HoodieWriteConfig.DEFAULT_MAX_LISTING_PARALLELISM.intValue())
+    val isPartitionedTable = _partitionSchema.fields.isEmpty
+    val serializableConf = new SerializableConfiguration(hadoopConf)
+    val partitionSchema = _partitionSchema
+    val timeZoneId = CaseInsensitiveMap(options)
+      .get(DateTimeUtils.TIMEZONE_OPTION)
+      .getOrElse(SQLConf.get.sessionLocalTimeZone)
+    // List files in all of the partition path.
+    val partition2Files =
+      spark.sparkContext.parallelize(partitionPaths, Math.min(partitionPaths.size, maxListParallelism))
+        .map { partitionPath =>
+          val partitionRow = if (isPartitionedTable) { // This is a none partitioned table
+            InternalRow.fromSeq(Seq(""))
+          } else {
+            val partitionSeqs = partitionPath.split("/")
+            assert(partitionSeqs.length == partitionSchema.size,
+              s"size of partition values[size is: ${partitionSeqs.size}, path " +
+                s"is: '$partitionPath'] is not equal to the size of" +
+                s" partition schema field[size is: ${partitionSchema.size}," +
+                s" fields is: ${partitionSchema.json}].")
+            // Append partition name to the partition value.
+            // e.g. convert "/xx/xx/2021/02" to "/xx/xx/year=2021/month=02"
+            val partitionWithName =
+            partitionSeqs.zip(partitionSchema).map {
+              case (partition, field) =>
+                if (partition.indexOf("=") == -1) {
+                  s"${field.name}=$partition"
+                } else {
+                  partition
+                }
+            }.mkString("/")
+            val pathWithPartitionName = new Path(basePath, partitionWithName)
+            val partitionDataTypes = partitionSchema.fields.map(f => f.name -> f.dataType).toMap
+            val partitionValues = PartitionUtils.parsePartition(pathWithPartitionName,
+              typeInference = false, Set(new Path(basePath)), partitionDataTypes,
+              DateTimeUtils.getTimeZone(timeZoneId))
+
+            // Convert partitionValues to InternalRow
+            partitionValues.map(_.literals.map(_.value))
+              .map(InternalRow.fromSeq)
+              .getOrElse(InternalRow.empty)
```

Comment:
To confirm my understanding, we are doing this part to be able to form `InternalRow` of partition values right, so it can evaluated with the predicates ? Re-using this logic from Spark is helping us avoid having to handle data types conversions ?

Reply:
Yes, you are right. I have Re used `PartitioningUtils.parsePartition` to convert the partition path to the `Literal` in the `partitionValues`. And then convert the `Literal` to `IntervalRow`.